test/gtest/common/test_perf.cc

```diff
@@ -37,18 +37,27 @@ void test_perf::rte_comm::push(const void *data, size_t size) {
 void test_perf::rte_comm::pop(void *data, size_t size,
                               void (*progress)(void *arg), void *arg) {
     bool done = false;
-    do {
+
+    for (;;) {
         pthread_mutex_lock(&m_mutex);
         if (m_queue.length() >= size) {
             memcpy(data, &m_queue[0], size);
             m_queue.erase(0, size);
             done = true;
         }
         pthread_mutex_unlock(&m_mutex);
-        if (!done) {
-            progress(arg);
+
+        if (done) {
+            break;
+        }
+
+        progress(arg);
+
+        /* Valgrind runs threads, but serializes them with one global lock */
+        if (ucs::running_inside_valgrind()) {
+            sched_yield();
```

Comment:
maybe it would be better to yield every N iterations to reduce syscall overhead?

Reply:
Double checked:
- current patch: 18.3sec 0.1 time variability
- yield 1/5: 18.9sec + time variability
- yield 1/10: 19.2sec + more time variability

We failed to pop, we progress locally (likely) as much as we can, so it seems better (under valgrind) to allow the other thread to progress in turn before trying to pop the result.